lib/Optimizer/CodeGen/ConvertToQIR.cpp

```diff
@@ -1077,6 +1077,164 @@ class MyToMz : public OpConversionPattern<quake::MyOp> {
   }
 };
 
+/// Convert quake.unitary to a call to __quantum__qis__unitary(data,
+/// controls, targets) or __quantum__qis__constant_unitary(real, imag, controls,
+/// targets)
+class UnitaryOpRewrite : public OpConversionPattern<quake::UnitaryOp> {
+public:
+  using OpConversionPattern::OpConversionPattern;
+
+  Value mapUnitaryDataToLLVMPointer(Location &loc,
+                                    ConversionPatternRewriter &rewriter,
+                                    ModuleOp parentModule,
+                                    std::int64_t numUnitaryElements,
+                                    std::string &operationName,
+                                    ArrayRef<double> data) const {
+    auto f64Ty = rewriter.getF64Type();
+    SmallVector<std::int64_t> shape{numUnitaryElements};
+    auto tensorTy = VectorType::get(shape, f64Ty);
+    auto insertPoint = rewriter.saveInsertionPoint();
+    rewriter.setInsertionPointToStart(parentModule.getBody());
+
+    LLVM::GlobalOp realGlobalOp;
+    if (auto existingArr =
+            parentModule.lookupSymbol<LLVM::GlobalOp>(operationName))
+      realGlobalOp = existingArr;
+    else
+      realGlobalOp = rewriter.create<LLVM::GlobalOp>(
+          loc, tensorTy, /*isConstant=*/true, LLVM::Linkage::Internal,
+          operationName, DenseFPElementsAttr::get(tensorTy, data),
+          /*alignment=*/0);
+
+    rewriter.restoreInsertionPoint(insertPoint);
+
+    Value realPartAddr = rewriter.create<LLVM::AddressOfOp>(
+        loc, cudaq::opt::factory::getPointerType(realGlobalOp.getType()),
+        realGlobalOp.getSymName());
+    Value zero =
+        rewriter.create<LLVM::ConstantOp>(loc, rewriter.getIntegerType(64), 0);
+    return rewriter.create<LLVM::GEPOp>(loc, LLVM::LLVMPointerType::get(f64Ty),
+                                        realPartAddr, ValueRange{zero, zero});
+  }
+
+  Value getQubitArray(Location &loc, ConversionPatternRewriter &rewriter,
+                      ValueRange qubits, FlatSymbolRefAttr &symbolRef,
+                      FlatSymbolRefAttr &getSymbolRef) const {
+
+    // Create (qubit...) -> Array* control array
+    auto ctx = rewriter.getContext();
+    Value eight = rewriter.create<arith::ConstantIntOp>(loc, 8, 32);
+    auto qirArrayTy = cudaq::opt::getArrayType(ctx);
+    auto i8PtrTy =
+        cudaq::opt::factory::getPointerType(IntegerType::get(ctx, 8));
+    Value numQubits =
+        rewriter.create<arith::ConstantIntOp>(loc, qubits.size(), 64);
+
+    auto createQubitsCall = rewriter.create<LLVM::CallOp>(
+        loc, qirArrayTy, symbolRef, ArrayRef<Value>{eight, numQubits});
+    auto controlsArr = createQubitsCall.getResult();
+    for (std::size_t i = 0; auto control : qubits) {
```

Comment:
```suggestion
    for (auto iter : llvm::enumerate(qubits)) {
      auto i = iter.index();
      auto control = iter.value();
```
for consistency with the rest of the code.

Is the variable `qubits` actually required to be controls here?